ethpm/_utils/chains.py

```diff
@@ -0,0 +1,102 @@
+import re
+from typing import Tuple
+from urllib import parse
+
+from eth_utils import add_0x_prefix, remove_0x_prefix, to_hex
+from web3 import Web3
+
+from ethpm.typing import URI
+
+
+def get_genesis_block_hash(web3: Web3) -> str:
+    return web3.eth.getBlock(0)["hash"]
+
+
+BLOCK = "block"
+
+BIP122_URL_REGEX = (
+    "^"
+    "blockchain://"
+    "(?P<chain_id>[a-zA-Z0-9]{64})"
+    "/"
+    "(?P<resource_type>block|transaction)"
+    "/"
+    "(?P<resource_hash>[a-zA-Z0-9]{64})"
+    "$"
+)
+
+
+def is_BIP122_uri(value: URI) -> bool:
+    return bool(re.match(BIP122_URL_REGEX, value))
+
+
+def parse_BIP122_uri(blockchain_uri: URI) -> Tuple[str, str, str]:
+    match = re.match(BIP122_URL_REGEX, blockchain_uri)
+    if match is None:
+        raise ValueError(f"Invalid URI format: '{blockchain_uri}'")
+    chain_id, resource_type, resource_hash = match.groups()
+    return (add_0x_prefix(chain_id), resource_type, add_0x_prefix(resource_hash))
+
+
+def is_BIP122_block_uri(value: URI) -> bool:
+    if not is_BIP122_uri(value):
+        return False
+    _, resource_type, _ = parse_BIP122_uri(value)
+    return resource_type == BLOCK
+
+
+BLOCK_OR_TRANSACTION_HASH_REGEX = "^(?:0x)?[a-zA-Z0-9]{64}$"
+
+
+def is_block_or_transaction_hash(value: str) -> bool:
+    return bool(re.match(BLOCK_OR_TRANSACTION_HASH_REGEX, value))
+
+
+def create_BIP122_uri(
+    chain_id: str, resource_type: str, resource_identifier: str
+) -> URI:
+    """
+    See: https://github.com/bitcoin/bips/blob/master/bip-0122.mediawiki
+    """
+    if resource_type != BLOCK:
+        raise ValueError("Invalid resource_type.  Must be one of 'block'")
+    elif not is_block_or_transaction_hash(resource_identifier):
+        raise ValueError(
+            "Invalid resource_identifier.  Must be a hex encoded 32 byte value"
+        )
+    elif not is_block_or_transaction_hash(chain_id):
+        raise ValueError("Invalid chain_id.  Must be a hex encoded 32 byte value")
```

Comment:
are these functions all just copy/pasted or do they need to be reviewed?

Reply:
everything's copy/pasted as far as what's inside a function body